File: serve.py
#!/bin/python3

import time
from concurrent import futures

import grpc

# Generated classes from the .proto file
import hello_pb2
import hello_pb2_grpc


# Implementation of the LandscapeHostAgent service
class GreetingsServer(hello_pb2_grpc.GreetingsServicer):

    def Connect(self, request_iterator, context):
        # Process the stream of Hello messages
        for hello in request_iterator:
            print("Received:", hello.data)
            yield hello_pb2.World(data=f"Hello back from the Python server! (you said: {hello.data})")

            # # Send a response for each message
            # while True:
            #     time.sleep(5)
                # yield hello_pb2.World(data=f"PING from Python server")

def serve_options():
    # return [
    #      ("grpc.keepalive_time_ms", 5000),
    #      ("grpc.keepalive_timeout_ms", 3000),
    # ]
    return  [
        ("grpc.max_connection_age_ms", 10_000),                                          

        # ("grpc.keepalive_time_ms", 30_000),
        # ("grpc.keepalive_timeout_ms", 10_000),
        # ("grpc.http2.min_ping_interval_without_data_ms", 5_000),
        # ("grpc.max_connection_idle_ms", 10_000),
        
        ("grpc.max_connection_age_grace_ms", 5_000),
        # ("grpc.http2.max_pings_without_data", 5),
           
        # ("grpc.keepalive_permit_without_calls", 1),
        # ("grpc.client_idle_timeout_ms", 15_000),
    ]

def serve():
    # Create a gRPC server
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10), options=serve_options())
    # Register the servicer with the server
    hello_pb2_grpc.add_GreetingsServicer_to_server(
        GreetingsServer(), server
    )
    # Bind the server to the port
    port = "8080"
    server.add_insecure_port(f"localhost:{port}")  # You can modify the port as needed
    server.start()
    print(f"Server started, listening on port :{port}")
    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        # server.stop(0)
        pass


if __name__ == "__main__":
    serve()

File: serve.go
package main

import (
	"fmt"
	"log"
	"net"
	"os"
	"time"

	helloworld "example.com/hello"
	"google.golang.org/grpc"
	"google.golang.org/grpc/keepalive"
)

// Define your gRPC service by implementing the generated interface
type GreetingsService struct {
	helloworld.UnimplementedGreetingsServer
	start time.Time
}

// Implement the methods defined in the service interface
func (s GreetingsService) Connect(stream helloworld.Greetings_ConnectServer) error {
	l := logger{time.Now()}

	msg, err := stream.Recv()
	if err != nil {
		l.Log("Could not receive: %v", err)
		return err
	}
	l.Log("Received message: %s", msg.GetData())

	if err := stream.Send(&helloworld.World{
		Data: fmt.Sprintf("Hello from Go server! (you sent: %s)", msg.GetData()),
	}); err != nil {
		l.Log("Could not send: %v", err)
		return err
	}

	// Implement your logic here
	for {
		time.Sleep(5 * time.Second)
		response := &helloworld.World{
			Data: "Hello from Go server",
		}
		if err := stream.Send(response); err != nil {
			l.Log("Could not send: %v", err)
			return err
		}
	}
}

type logger struct {
	start time.Time
}

func (l logger) Log(msg string, args ...interface{}) {
	t := int(time.Since(l.start) / time.Second)
	if _, err := fmt.Printf("[%ds] %s\n", t, fmt.Sprintf(msg, args...)); err != nil {
		panic(err)
	}
}

func (l logger) LogFatal(msg string, args ...interface{}) {
	l.Log(msg, args...)
	os.Exit(1)
}

func main() {
	// Create a gRPC server
	server := grpc.NewServer(grpc.KeepaliveEnforcementPolicy(enfPolicy), grpc.KeepaliveParams(serverParams))

	// Register your service with the server
	GreetingsService := &GreetingsService{
		start: time.Now(),
	}
	helloworld.RegisterGreetingsServer(server, GreetingsService)

	// Start listening on a specific port
	listener, err := net.Listen("tcp", ":8080")
	if err != nil {
		log.Fatalf("Failed to listen: %v", err)
	}

	log.Printf("Server started at %s", listener.Addr())

	// Start serving requests
	if err := server.Serve(listener); err != nil {
		log.Fatalf("Failed to serve: %v", err)
	}
}

var enfPolicy = keepalive.EnforcementPolicy{
	// MinTime:             5 * time.Second,
	// PermitWithoutStream: true,
}

var serverParams = keepalive.ServerParameters{
	// MaxConnectionIdle is a duration for the amount of time after which an
	// idle connection would be closed by sending a GoAway. Idleness duration is
	// defined since the most recent time the number of outstanding RPCs became
	// zero or the connection establishment.
	MaxConnectionIdle: 5 * time.Second,

	// MaxConnectionAge is a duration for the maximum amount of time a
	// connection may exist before it will be closed by sending a GoAway. A
	// random jitter of +/-10% will be added to MaxConnectionAge to spread out
	// connection storms.
	//MaxConnectionAge: 10 * time.Second, // The current default value is infinity.

	// MaxConnectionAgeGrace is an additive period after MaxConnectionAge after
	// which the connection will be forcibly closed.
	//MaxConnectionAgeGrace: 5 * time.Second, // The current default value is infinity.

	// After a duration of this time if the server doesn't see any activity it
	// pings the client to see if the transport is still alive.
	// If set below 1s, a minimum value of 1s will be used instead.
	// Time time.Duration // The current default value is 2 hours.

	// After having pinged for keepalive check, the server waits for a duration
	// of Timeout and if no activity is seen even after that the connection is
	// closed.
	// Timeout: time.Second,
}
